LSPS2/README.md

```diff
@@ -0,0 +1,736 @@
+  LSP SHOULD error the client request in the next step).
+  * Unsigned integer division cannot overflow.
+  * Addition overflow can be detected by comparing the sum to an
+    addend:
+    if the sum is less than one addend, overflow has occurred and
+    the flow must be aborted.
+  * Multiplication overflow can be detected by checking that the
+    first factor is nonzero, and dividing the product by the first
+    factor does not result in the second factor.
+  * The check MUST be done on EACH addition and multiplication
+    operation, including those with constants.
+
+The C routine `compute_opening_fee` below performs the computation,
+and indicates an error (returns `-1`) and sets `errno` if the
+computation fails, otherwise returns `0` on success.
+
+```C
+#include<errno.h>
+#include<stdint.h>
+
+static inline int
+ov_add (uint64_t *sum, uint64_t a, uint64_t b) {
+  *sum = a + b;
+  if (*sum < a)
+  {
+    errno = EDOM;
+    return -1;
+  }
+  return 0;
+}
+static inline int
+ov_mul (uint64_t *prod, uint64_t a, uint64_t b) {
+  *prod = a * b;
+  if ((a != 0) && ((*prod / a) != b))
+  {
+    errno = EDOM;
+    return -1;
+  }
+  return 0;
+}
+
+int
+compute_opening_fee(uint64_t* opening_fee, uint64_t payment_size,
+		    uinr64_t opening_fee_base, uint64_t opening_fee_proportional) {
+  uint64_t tmp;
+  if (0 != ov_mul (&tmp, payment_size, opening_fee_proportional))
+    return -1;
+  if (0 != ov_add (&tmp, tmp, 999999))
+    return -1;
+  /* C specifies division rounds towards 0.  */
+  tmp = tmp / 1000000;
+  if (0 != ov_add (&tmp, tmp, opening_fee_base))
+    return -1;
+  *opening_fee = tmp;
+  return 0;
+}
+```
+
+> **Rationale** The computation is described in excruciating detail
+> in order to ensure that both LSP and client agree on the result
+> of computation of the `opening_fee`.
+> In C, signed integer overflow is undefined behavior, and even if
+> neither LSP nor client is written in C, they may be written in a
+> language whose implementation is an interpreter written in C or
+> using C libraries, which may also trigger the same undefined
+> behavior on signed integer overflow, thus unsigned integers are
+> specified, which, in C, are specifically modulo 2 to the number
+> of bits.
+> Arithmetic overflow is problematic and may lead to unexpectedly
+> low `opening_fee` charges for high `payment_size` and high
+> `opening_fee_params.proportional`, thus must be specifically
+> detected and cause an abort in this flow rather than continue
+> and cause problems for the LSP.
+> Most modern microprocessors have dedicated instructions to detect
+> overflow, and most modern compilers can recognize the above code
+> and optimize those down to these dedicated instructions at high
+> optimization settings; in particular, the division in the
+> overflow-detecting multiplication routine is optimized away
+> and replaced with a simple overflow-flag check.
+
+### 2.  Request JIT Channel
```

Comment:
One potentially missing detail in this section. The client will be connected to the LSP with a specific identity pubkey. The LSP will correlate the SCID with that identity pubkey. This is how the LSP knows that that client is connected when the htlc arrives. So the client must be connected with the same identity pubkey throughout the flow.